pkg/controller/direct/alloydb/mapping.go

```diff
@@ -0,0 +1,231 @@
+	if in == nil {
+		return nil
+	}
+	out := &api.AutomatedBackupPolicy{
+		BackupWindow:           ValueOf(in.BackupWindow),
+		Enabled:                ValueOf(in.Enabled),
+		EncryptionConfig:       EncryptionConfig_KRMToApi(ctx, in.EncryptionConfig),
+		Labels:                 in.Labels,
+		Location:               ValueOf(in.Location),
+		QuantityBasedRetention: QuantityBasedRetention_KRMToApi(ctx, in.QuantityBasedRetention),
+		TimeBasedRetention:     TimeBasedRetention_KRMToApi(ctx, in.TimeBasedRetention),
+		WeeklySchedule:         WeeklySchedule_KRMToApi(ctx, in.WeeklySchedule),
+	}
+	return out
+}
+
+func WeeklySchedule_KRMToApi(ctx *MapContext, in *krm.ClusterWeeklySchedule) *api.WeeklySchedule {
+	if in == nil {
+		return nil
+	}
+	out := &api.WeeklySchedule{
+		DaysOfWeek: in.DaysOfWeek,
+		StartTimes: StartTimes_KRMToApi(ctx, in.StartTimes),
+	}
+
+	return out
+}
+
+func StartTimes_KRMToApi(ctx *MapContext, in []krm.ClusterStartTimes) []*api.GoogleTypeTimeOfDay {
+	out := make([]*api.GoogleTypeTimeOfDay, len(in))
+	for i, v := range in {
+		out[i] = Time_KRMToApi(ctx, LazyPtr(v))
+	}
+	return out
+}
+
+func Time_KRMToApi(ctx *MapContext, in *krm.ClusterStartTimes) *api.GoogleTypeTimeOfDay {
+	if in == nil {
+		return nil
+	}
+	out := &api.GoogleTypeTimeOfDay{
+		Hours:   int64(ValueOf(in.Hours)),
+		Minutes: int64(ValueOf(in.Minutes)),
+		Nanos:   int64(ValueOf(in.Nanos)),
+		Seconds: int64(ValueOf(in.Seconds)),
+	}
+	return out
+}
+
+func EncryptionConfig_KRMToApi(ctx *MapContext, in *krm.ClusterEncryptionConfig) *api.EncryptionConfig {
+	if in == nil {
+		return nil
+	}
+	out := &api.EncryptionConfig{
+		KmsKeyName: in.KmsKeyNameRef.External,
+	}
+	return out
+}
+
+func TimeBasedRetention_KRMToApi(ctx *MapContext, in *krm.ClusterTimeBasedRetention) *api.TimeBasedRetention {
+	if in == nil {
+		return nil
+	}
+	out := &api.TimeBasedRetention{
+		RetentionPeriod: ValueOf(in.RetentionPeriod),
+	}
+
+	return out
+}
+
+func QuantityBasedRetention_KRMToApi(ctx *MapContext, in *krm.ClusterQuantityBasedRetention) *api.QuantityBasedRetention {
+	if in == nil {
+		return nil
+	}
+	out := &api.QuantityBasedRetention{
+		Count: int64(ValueOf(in.Count)),
+	}
+
+	return out
+}
+
+func ContinuousBackupConfig_KRMToApi(ctx *MapContext, in *krm.ClusterContinuousBackupConfig) *api.ContinuousBackupConfig {
+	if in == nil {
+		return nil
+	}
+	out := &api.ContinuousBackupConfig{
+		Enabled:            ValueOf(in.Enabled),
+		EncryptionConfig:   EncryptionConfig_KRMToApi(ctx, in.EncryptionConfig),
+		RecoveryWindowDays: int64(ValueOf(in.RecoveryWindowDays)),
+	}
+
+	return out
+}
+
+func InitialUser_KRMToApi(ctx *MapContext, in *krm.ClusterInitialUser) *api.UserPassword {
+	if in == nil {
+		return nil
+	}
+	out := &api.UserPassword{
+		User: ValueOf(in.User),
+	}
+	if in.Password.Value != nil {
+		out.Password = ValueOf(in.Password.Value)
+	} else {
+		out.Password = in.Password.ValueFrom.SecretKeyRef.Key
```

Comment:
Is this right?  You can add errors to `ctx *MapContext` if that's more appropriate.

One thing I'm not certain about is when we should resolve these i.e. whether we resolve them on demand during the mapping, or whether we try to resolve them all up-front and (here) replace ValueFrom with  Value.  For ProjectRefs, we would replace Name with External. etc